packages/flutter/lib/src/material/text_field.dart

```diff
@@ -601,6 +613,15 @@ class _TextFieldState extends State<TextField> with AutomaticKeepAliveClientMixi
     if (wasEnabled && !isEnabled) {
       _effectiveFocusNode.unfocus();
     }
+    if (_effectiveFocusNode.hasFocus && widget.readOnly != oldWidget.readOnly) {
```

Comment:
this logic deal with collapse selection handle during readonly <---> non readonly transition.

To make it possible _editableText.showHandles() will need to remove some assert (see editable_text.dart line 1452)
The assert is fine before because text_field used to only call the function during handle selection change. At the time selectionoverlay is guaranteed to exist. However, this is not the case here, and TextField does not have knowledget about selectionoverlay.

The other part I need to change is text_selection.dart line 358. showHandles will now be called during build phase, I have to deal with that case.

I make this compromise since I can't think of a better way. The selection handle completely handle by the TextField side instead of Editabletext due to we need different behavior on CupertinoTextField and Textfield.

Reply:
@justinmc

Reply:
also added a test 'Dynamically switching between read only and not read only should hide or show collapse cursor'

Reply:
https://github.com/flutter/flutter/commit/bcae3563d4
this is the pr that exposed showhandles, and I think there is no easy way to avoid it because we want different showhandles behavior on TextField and CupertinoTextfield. Unless we can move the entire selectionOverlay out of editable text, which is a heavy lift but probably the right approach.

This logic does not need to be in CupertinoTextField as it already does not show the handles when it is collapse

Reply:
@HansMuller might have more context on this?

Reply:
I didn't review https://github.com/flutter/flutter/pull/29683 but it does seem strange that we're adding a `showCursor` property where we already have `showToolbar()` and `hideToolbar()` methods. The toolbar methods seem out of place. 

This PR introduces some backwards-incompatible changes anyway ... maybe we should replace showToolbar and hideToolbar with a showToolbar property? The toolbar methods have only been around for two weeks.

Reply:
showcursor and showhandles are two different things.
![Screen Shot 2019-05-29 at 9 41 42 AM](https://user-images.githubusercontent.com/47866232/58575150-5fc20700-81f6-11e9-8646-49c15cdb6d98.png)